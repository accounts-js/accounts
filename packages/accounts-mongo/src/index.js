// @flow

import { get } from 'lodash';
import { encryption } from '@accounts/server';
import type {
  CreateUserType,
  UserObjectType,
  SessionType,
} from '@accounts/common';
import { ObjectID } from 'mongodb';

export type MongoOptionsType = {
  collectionName: string,
  sessionCollectionName: string,
  timestamps: {
    createdAt: string,
    updatedAt: string,
  }
};

export type MongoUserObjectType = {
  username?: string,
  profile?: Object,
  services: {
    password?: {
      bcrypt: string,
    },
  },
  emails?: [{
    address: string,
    verified: boolean,
  }],
};

const toMongoID = (objectId) => {
  if (typeof objectId === 'string') {
    return new ObjectID(objectId);
  }

  return objectId;
};

class Mongo {
  options: MongoOptionsType;
  // TODO definition for mongodb connection object
  db: any;
  collection: any;
  sessionCollection: any;

  constructor(db: any, options: MongoOptionsType) {
    const defaultOptions = {
      collectionName: 'users',
      sessionCollectionName: 'sessions',
      timestamps: {
        createdAt: 'createdAt',
        updatedAt: 'updatedAt',
      },
    };
    this.options = { ...defaultOptions, ...options };
    if (get(db, 'constructor.define.name') !== 'Db') {
      throw new Error('A valid database connection object is required');
    }
    this.db = db;
    this.collection = this.db.collection(this.options.collectionName);
    this.sessionCollection = this.db.collection(this.options.sessionCollectionName);
  }

  async setupIndexes(): Promise<void> {
    await this.collection.createIndex('username', { unique: 1, sparse: 1 });
    await this.collection.createIndex('emails.address', { unique: 1, sparse: 1 });
  }

  async createUser(options: CreateUserType): Promise<string> {
    const user: MongoUserObjectType = {
      services: {},
      profile: {},
      [this.options.timestamps.createdAt]: Date.now(),
      [this.options.timestamps.updatedAt]: Date.now(),
    };
    if (options.password) {
      user.services.password = { bcrypt: await encryption.hashPassword(options.password) };
    }
    if (options.username) {
      user.username = options.username;
    }
    if (options.email) {
      user.emails = [{ address: options.email.toLowerCase(), verified: false }];
    }
    if (options.profile) {
      user.profile = options.profile;
    }
    const ret = await this.collection.insertOne(user);
    return ret.ops[0]._id;
  }

  async findUserById(userId: string): Promise<?UserObjectType> {
    const user = await this.collection.findOne({ _id: toMongoID(userId) });
    if (user) {
      user.id = user._id;
    }
    return user;
  }

  async findUserByEmail(email: string): Promise<?UserObjectType> {
    const user = await this.collection.findOne({ 'emails.address': email.toLowerCase() });
    if (user) {
      user.id = user._id;
    }
    return user;
  }

  async findUserByUsername(username: string): Promise<?UserObjectType> {
    const user = await this.collection.findOne({ username });
    if (user) {
      user.id = user._id;
    }
    return user;
  }

  async findPasswordHash(userId: string): Promise<?string> {
    const user = await this.findUserById(toMongoID(userId));
    if (user) {
      return user.services.password.bcrypt;
    }
    return null;
  }

  async addEmail(userId: string, newEmail: string, verified: boolean): Promise<void> {
    const ret = await this.collection.update({ _id: toMongoID(userId) }, {
      $addToSet: {
        emails: {
          address: newEmail.toLowerCase(),
          verified,
        },
      },
      $set: { [this.options.timestamps.updatedAt]: Date.now() },
    });
    if (ret.result.nModified === 0) {
      throw new Error('User not found');
    }
  }

  async removeEmail(userId: string, email: string): Promise<void> {
    const ret = await this.collection.update({ _id: toMongoID(userId) }, {
      $pull: { emails: { address: email.toLowerCase() } },
      $set: { [this.options.timestamps.updatedAt]: Date.now() },
    });
    if (ret.result.nModified === 0) {
      throw new Error('User not found');
    }
  }

  async setUsername(userId: string, newUsername: string): Promise<void> {
    const ret = await this.collection.update({ _id: toMongoID(userId) }, {
      $set: {
        username: newUsername,
        [this.options.timestamps.updatedAt]: Date.now(),
      },
    });
    if (ret.result.nModified === 0) {
      throw new Error('User not found');
    }
  }

  async setPasssword(userId: string, newPassword: string): Promise<void> {
    const ret = await this.collection.update({ _id: toMongoID(userId) }, {
      $set: {
        'services.password.bcrypt': await encryption.hashPassword(newPassword),
        [this.options.timestamps.updatedAt]: Date.now(),
      },
    });
    if (ret.result.nModified === 0) {
      throw new Error('User not found');
    }
  }

  async setProfile(userId: string, profile: Object): Promise<Object> {
    await this.collection.update({ _id: toMongoID(userId) }, {
      $set: {
        profile,
        [this.options.timestamps.updatedAt]: Date.now(),
      },
    });
    const user = await this.findUserById(userId);
    return user && user.profile;
  }

  async createSession(userId: string, ip: string, userAgent: string): Promise<string> {
    const ret = await this.sessionCollection.insertOne({
      userId,
      userAgent,
      ip,
      valid: true,
      [this.options.timestamps.createdAt]: Date.now(),
      [this.options.timestamps.updatedAt]: Date.now(),
    });
    return ret.ops[0]._id;
  }

  async updateSession(sessionId: string, ip: string, userAgent: string): Promise<void> {
    await this.sessionCollection.update({ _id: toMongoID(sessionId) }, {
      $set: {
        ip,
        userAgent,
        [this.options.timestamps.updatedAt]: Date.now(),
      },
    });
  }

  async invalidateSession(sessionId: string): Promise<void> {
    await this.sessionCollection.update({ _id: toMongoID(sessionId) }, {
      $set: {
        valid: false,
        [this.options.timestamps.updatedAt]: Date.now(),
      },
    });
  }

  findSessionById(sessionId: string): Promise<?SessionType> {
    return this.sessionCollection.findOne({ _id: toMongoID(sessionId) });
  }
}

export default Mongo;
