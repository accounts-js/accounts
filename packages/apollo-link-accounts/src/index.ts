import { setContext } from 'apollo-link-context';
import { ApolloLink } from 'apollo-link';
import { AccountsClient } from '@accounts/client';

export const accountsLink = (accountsClient: AccountsClient): ApolloLink => {
  return setContext(async (_, { headers }) => {
    const tokens = await accountsClient.refreshSession();
    return {
      headers: {
        ...headers,
        'accounts-access-token': tokens ? tokens.accessToken : null,
      },
    };
  });
};
