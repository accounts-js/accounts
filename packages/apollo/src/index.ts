import { accountsLink } from '@accounts/apollo-link';
import { AccountsClientOptions, AccountsClient } from '@accounts/client';
import AccountsGraphQLClient from '@accounts/graphql-client';
import { ApolloClient } from 'apollo-client';
import { ApolloLink } from 'apollo-link';
import { onError } from 'apollo-link-error';
import { HttpLink } from 'apollo-link-http';
import { InMemoryCache } from 'apollo-cache-inmemory';

export interface AccountsApolloOptions extends AccountsClientOptions {
  uri: string;
}

export const accountsApollo = (options: AccountsApolloOptions) => {
  const accountsApolloClient = new ApolloClient({
    cache: new InMemoryCache(),
    link: ApolloLink.from([
      onError(({ graphQLErrors, networkError }) => {
        if (graphQLErrors) {
          graphQLErrors.map(({ message, locations, path }) =>
            console.log(
              `[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`
            )
          );
        }
        if (networkError) {
          console.log(`[Network error]: ${networkError}`);
        }
      }),
      new HttpLink({
        uri: options.uri,
      }),
    ]),
  });

  const accountsClient = new AccountsClient(
    options,
    new AccountsGraphQLClient({
      graphQLClient: accountsApolloClient,
    })
  );

  let accountsPassword;
  if (require.resolve('@accounts/client-password')) {
    const AccountsClientPassword = require('@accounts/client-password').AccountsClientPassword; // tslint:disable-line no-implicit-dependencies
    accountsPassword = new AccountsClientPassword(accountsClient);
  }

  return {
    accountsClient,
    accountsApolloClient,
    accountsLink: accountsLink(accountsClient),
    accountsPassword,
  };
};

export default accountsApollo;
