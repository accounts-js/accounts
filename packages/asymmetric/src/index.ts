import AccountsAsymmetric from './accounts-asymmetric';

export default AccountsAsymmetric;
export { AccountsAsymmetric };
