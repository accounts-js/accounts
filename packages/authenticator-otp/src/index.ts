import {
  DatabaseInterface,
  AuthenticatorService,
  Authenticator,
  MfaChallenge,
} from '@accounts/types';
import { AccountsServer, generateRandomToken } from '@accounts/server';
import { authenticator as optlibAuthenticator } from 'otplib';

// TODO remove the url to be returned
// - should be done client side
// - update documentation to see how to use it

interface DbAuthenticatorOtp extends Authenticator {
  secret: string;
}

export interface AuthenticatorOtpOptions {
  /**
   * Two factor app name that will be displayed inside the user authenticator app.
   */
  appName?: string;

  /**
   * Two factor user name that will be displayed inside the user authenticator app,
   * usually a name, email etc..
   * Will be called every time a user register a new device.
   * That way you can display something like "Github (leo@accountsjs.com)" in the authenticator app.
   */
  userName?: (userId: string) => Promise<string> | string;
}

const defaultOptions = {
  appName: 'accounts-js',
};

export class AuthenticatorOtp implements AuthenticatorService {
  public authenticatorName = 'otp';
  public server!: AccountsServer;

  private options: AuthenticatorOtpOptions & typeof defaultOptions;
  private db!: DatabaseInterface;

  constructor(options: AuthenticatorOtpOptions = {}) {
    this.options = { ...defaultOptions, ...options };
  }

  public setStore(store: DatabaseInterface) {
    this.db = store;
  }

  /**
   * @description Start the association of a new OTP device
   */
  public async associate(
    userIdOrMfaChallenge: string | MfaChallenge
  ): Promise<{ id: string; mfaToken: string; secret: string; otpauthUri: string }> {
    const userId =
      typeof userIdOrMfaChallenge === 'string' ? userIdOrMfaChallenge : userIdOrMfaChallenge.userId;
    const mfaChallenge = typeof userIdOrMfaChallenge === 'string' ? null : userIdOrMfaChallenge;

    const secret = optlibAuthenticator.generateSecret();
    const userName = this.options.userName ? await this.options.userName(userId) : userId;
    const otpauthUri = optlibAuthenticator.keyuri(userName, this.options.appName, secret);

    const authenticatorId = await this.db.createAuthenticator({
      type: this.authenticatorName,
      userId,
      secret,
      active: false,
    });

    let mfaToken: string;
    if (mfaChallenge) {
      mfaToken = mfaChallenge.token;
      await this.db.updateMfaChallenge(mfaChallenge.id, {
        authenticatorId,
      });
    } else {
      // We create a new challenge for the authenticator so it can be verified later
      mfaToken = generateRandomToken();
      await this.db.createMfaChallenge({
        userId,
        authenticatorId,
        token: mfaToken,
        scope: 'associate',
      });
    }

    return {
      id: authenticatorId,
      mfaToken,
      secret,
      otpauthUri,
    };
  }

  /**
   * @description Verify that the code provided by the user is valid
   */
  public async authenticate(
    mfaChallenge: MfaChallenge,
    authenticator: DbAuthenticatorOtp,
    { code }: { code?: string }
  ): Promise<boolean> {
    if (!code) {
      throw new Error('Code required');
    }

    return optlibAuthenticator.check(code, authenticator.secret);
  }

  /**
   * @description Remove the sensitive fields from the database authenticator. The object
   * returned by this function can be exposed to the user safely.
   */
  public sanitize(authenticator: DbAuthenticatorOtp): Authenticator {
    // The secret key should never be exposed to the user after the authenticator is linked
    const {
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      secret,
      ...safeAuthenticator
    } = authenticator;
    return safeAuthenticator;
  }
}
