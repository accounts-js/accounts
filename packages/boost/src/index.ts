import { DatabaseManager } from '@accounts/database-manager';
import {
  createAccountsCoreModule,
  createAccountsPasswordModule,
  authenticated,
  context,
} from '@accounts/graphql-api';
import { AccountsServer, AccountsServerOptions } from '@accounts/server';
import { AuthenticationService } from '@accounts/types';
import { ApolloServer } from 'apollo-server';
import { verify } from 'jsonwebtoken';
import { get, isString, merge } from 'lodash';
import { Application, createApplication } from 'graphql-modules';

export { createAccountsCoreModule, createAccountsPasswordModule };

export { AccountsServerOptions };

export { authenticated };

export interface AccountsBoostOptions extends AccountsServerOptions {
  storage?: {
    uri?: string;
    name?: string;
  };
  micro?: boolean;
}

const defaultAccountsBoostOptions = {
  storage: {
    uri: 'mongodb://localhost:27017',
    name: 'accounts-js',
  },
};

const requirePackage = (packageName: string) => {
  if (require.resolve(packageName)) {
    return require(packageName);
  }
};

export const accountsBoost = async (userOptions?: AccountsBoostOptions): Promise<AccountsBoost> => {
  const options = merge({}, defaultAccountsBoostOptions, userOptions);

  const databasePackages = {
    ['@accounts/mongo']: async (requiredPackage: any): Promise<any> => {
      // The `@accounts/mongo` package comes with the `mongodb` driver
      // eslint-disable-next-line
      const mongodb = require('mongodb');

      const mongoClient = (await mongodb.MongoClient.connect(get(options, 'storage.uri'))).db(
        get(options, 'storage.name')
      );

      return new requiredPackage.Mongo(mongoClient, options);
    },
  };

  const storage = await Object.keys(databasePackages).reduce(
    async (res, packageName: string): Promise<any> => {
      const requiredPackage = requirePackage(packageName);
      if (requiredPackage) {
        return (databasePackages as any)[packageName](requiredPackage);
      }
      return res;
    },
    Promise.resolve<any>([])
  );

  if (!storage) {
    throw new Error('A database package could not be loaded. Did you install one?');
  }

  options.db = new DatabaseManager({
    userStorage: storage,
    sessionStorage: storage,
  });

  const servicePackages = {
    ['@accounts/password']: async (requiredPackage: any): Promise<any> => {
      const AccountsPassword = requiredPackage.default;
      return new AccountsPassword(get(options, ['services', 'password']));
    },
  };

  const services = await Object.keys(servicePackages).reduce(
    async (res, packageName: string): Promise<any> => {
      const requiredPackage = requirePackage(packageName);
      if (requiredPackage) {
        const service = await (servicePackages as any)[packageName](requiredPackage);
        return {
          ...res,
          [service.serviceName]: service,
        };
      }
      return res;
    },
    Promise.resolve({})
  );

  // eslint-disable-next-line
  return new AccountsBoost(options, services);
};

export interface AccountsBoostListenOptions {
  port?: number;
}

const defaultAccountsBoostListenOptions: AccountsBoostListenOptions = {
  port: 4003,
};

export class AccountsBoost {
  public accountsServer: AccountsServer;
  public apolloServer: ApolloServer;
  public accountsGraphQL: Application;
  private options: AccountsBoostOptions;

  constructor(options: AccountsBoostOptions, services: { [key: string]: AuthenticationService }) {
    this.accountsServer = new AccountsServer(options, services);
    this.options = options;
    this.accountsGraphQL = createApplication({
      modules: [createAccountsCoreModule({ accountsServer: this.accountsServer })],
    });

    const { schema } = this.accountsGraphQL;

    this.apolloServer = new ApolloServer({
      schema,
      context: ({ req, connection }) => {
        return context(
          { req, connection },
          {
            accountsServer: this.accountsServer,
          }
        );
      },
      //schemaDirectives
    });
  }

  public async listen(options?: AccountsBoostListenOptions): Promise<any> {
    const res = await this.apolloServer.listen(
      merge({}, defaultAccountsBoostListenOptions, options)
    );

    console.log(`Accounts GraphQL server running at ${res.url}`);

    return res;
  }

  public graphql(): Application {
    // Cache `this.accountsGraphQL` to avoid regenerating the schema if the user calls `accountsBoost.graphql()` multple times.
    if (this.accountsGraphQL) {
      return this.accountsGraphQL;
    }

    if (this.options.micro) {
      this.accountsServer.resumeSession = async (accessToken: string) => {
        let decoded: any;

        if (!isString(accessToken)) {
          throw new Error('An access token is required');
        }

        try {
          const secretOrPublicKey =
            typeof this.options.tokenSecret === 'string'
              ? this.options.tokenSecret
              : this.options.tokenSecret.publicKey;
          decoded = verify(accessToken, secretOrPublicKey);
        } catch (err) {
          throw new Error('Access token is not valid');
        }

        return {
          id: decoded.data.userId,
        } as any;
      };
    }

    return this.accountsGraphQL;
  }
}

export default accountsBoost;
