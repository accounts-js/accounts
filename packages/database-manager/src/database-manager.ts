import { DatabaseInterface, DatabaseInterfaceSessions } from '@accounts/types';

import { Configuration } from './types/configuration';

export class DatabaseManager implements DatabaseInterface {
  private userStorage: DatabaseInterface;
  private sessionStorage: DatabaseInterface | DatabaseInterfaceSessions;

  constructor(configuration: Configuration) {
    this.validateConfiguration(configuration);
    this.userStorage = configuration.userStorage;
    this.sessionStorage = configuration.sessionStorage;
  }

  private validateConfiguration(configuration: Configuration): void {
    if (!configuration) {
      throw new Error(
        '[ Accounts - DatabaseManager ] configuration : A configuration object is required on DatabaseManager'
      );
    }
    if (!configuration.userStorage) {
      throw new Error(
        '[ Accounts - DatabaseManager ] configuration : A userStorage DatabaseInterface is required'
      );
    }
    if (!configuration.sessionStorage) {
      throw new Error(
        '[ Accounts - DatabaseManager ] configuration : A sessionStorage DatabaseInterface is required'
      );
    }
  }

  // Return the createUser function from the userStorage
  public get createUser(): DatabaseInterface['createUser'] {
    return this.userStorage.createUser.bind(this.userStorage);
  }

  // Return the findUserById function from the userStorage
  public get findUserById(): DatabaseInterface['findUserById'] {
    return this.userStorage.findUserById.bind(this.userStorage);
  }

  // Return the findUserByEmail function from the userStorage
  public get findUserByEmail(): DatabaseInterface['findUserByEmail'] {
    return this.userStorage.findUserByEmail.bind(this.userStorage);
  }

  // Return the findUserByUsername function from the userStorage
  public get findUserByUsername(): DatabaseInterface['findUserByUsername'] {
    return this.userStorage.findUserByUsername.bind(this.userStorage);
  }

  // Return the findPasswordHash function from the userStorage
  public get findPasswordHash(): DatabaseInterface['findPasswordHash'] {
    return this.userStorage.findPasswordHash.bind(this.userStorage);
  }

  // Return the findUserByEmailVerificationToken function from the userStorage
  public get findUserByEmailVerificationToken(): DatabaseInterface['findUserByEmailVerificationToken'] {
    return this.userStorage.findUserByEmailVerificationToken.bind(this.userStorage);
  }

  // Return the findUserByResetPasswordToken function from the userStorage
  public get findUserByResetPasswordToken(): DatabaseInterface['findUserByResetPasswordToken'] {
    return this.userStorage.findUserByResetPasswordToken.bind(this.userStorage);
  }

  // Return the findUserByServiceId function from the userStorage
  public get findUserByServiceId(): DatabaseInterface['findUserByServiceId'] {
    return this.userStorage.findUserByServiceId.bind(this.userStorage);
  }

  // Return the addEmail function from the userStorage
  public get addEmail(): DatabaseInterface['addEmail'] {
    return this.userStorage.addEmail.bind(this.userStorage);
  }

  // Return the removeEmail function from the userStorage
  public get removeEmail(): DatabaseInterface['removeEmail'] {
    return this.userStorage.removeEmail.bind(this.userStorage);
  }

  // Return the verifyEmail function from the userStorage
  public get verifyEmail(): DatabaseInterface['verifyEmail'] {
    return this.userStorage.verifyEmail.bind(this.userStorage);
  }

  // Return the setUsername function from the userStorage
  public get setUsername(): DatabaseInterface['setUsername'] {
    return this.userStorage.setUsername.bind(this.userStorage);
  }

  // Return the setPassword function from the userStorage
  public get setPassword(): DatabaseInterface['setPassword'] {
    return this.userStorage.setPassword.bind(this.userStorage);
  }

  // Return the setService function from the userStorage
  public get setService(): DatabaseInterface['setService'] {
    return this.userStorage.setService.bind(this.userStorage);
  }

  // Return the unsetService function from the userStorage
  public get unsetService(): DatabaseInterface['unsetService'] {
    return this.userStorage.unsetService.bind(this.userStorage);
  }

  // Return the createSession function from the sessionStorage
  public get createSession(): DatabaseInterface['createSession'] {
    return this.sessionStorage.createSession.bind(this.sessionStorage);
  }

  // Return the updateSession function from the sessionStorage
  public get updateSession(): DatabaseInterface['updateSession'] {
    return this.sessionStorage.updateSession.bind(this.sessionStorage);
  }

  // Return the invalidateSession function from the sessionStorage
  public get invalidateSession(): DatabaseInterface['invalidateSession'] {
    return this.sessionStorage.invalidateSession.bind(this.sessionStorage);
  }

  // Return the invalidateAllSessions function from the sessionStorage
  public get invalidateAllSessions(): DatabaseInterface['invalidateAllSessions'] {
    return this.sessionStorage.invalidateAllSessions.bind(this.sessionStorage);
  }

  // Return the removeAllResetPasswordTokens function from the sessionStorage
  public get removeAllResetPasswordTokens(): DatabaseInterface['removeAllResetPasswordTokens'] {
    return this.userStorage.removeAllResetPasswordTokens.bind(this.userStorage);
  }

  // Return the findSessionByToken function from the sessionStorage
  public get findSessionByToken(): DatabaseInterface['findSessionByToken'] {
    return this.sessionStorage.findSessionByToken.bind(this.sessionStorage);
  }

  // Return the findSessionById function from the sessionStorage
  public get findSessionById(): DatabaseInterface['findSessionById'] {
    return this.sessionStorage.findSessionById.bind(this.sessionStorage);
  }

  // Return the addEmailVerificationToken function from the userStorage
  public get addEmailVerificationToken(): DatabaseInterface['addEmailVerificationToken'] {
    return this.userStorage.addEmailVerificationToken.bind(this.userStorage);
  }

  // Return the addResetPasswordToken function from the userStorage
  public get addResetPasswordToken(): DatabaseInterface['addResetPasswordToken'] {
    return this.userStorage.addResetPasswordToken.bind(this.userStorage);
  }

  // Return the setUserDeactivated function from the userStorage
  public get setUserDeactivated(): DatabaseInterface['setUserDeactivated'] {
    return this.userStorage.setUserDeactivated.bind(this.userStorage);
  }

  // Return the findUserByLoginToken function from the userStorage
  public get findUserByLoginToken(): DatabaseInterface['findUserByLoginToken'] {
    return this.userStorage.findUserByLoginToken.bind(this.userStorage);
  }

  // Return the addLoginToken function from the userStorage
  public get addLoginToken(): DatabaseInterface['addLoginToken'] {
    return this.userStorage.addLoginToken.bind(this.userStorage);
  }

  // Return the addLoginToken function from the userStorage
  public get removeAllLoginTokens(): DatabaseInterface['removeAllLoginTokens'] {
    return this.userStorage.removeAllLoginTokens.bind(this.userStorage);
  }
}
