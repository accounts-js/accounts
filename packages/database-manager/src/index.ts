export { DatabaseManager } from './database-manager';
