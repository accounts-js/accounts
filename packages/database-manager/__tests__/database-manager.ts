import { DatabaseManager } from '../src';

// eslint-disable-next-line jest/no-export
export default class Database {
  public name: any;

  constructor(name: any) {
    this.name = name;
  }

  public createUser() {
    return this.name;
  }

  public findUserById() {
    return this.name;
  }

  public findUserByEmail() {
    return this.name;
  }

  public findUserByUsername() {
    return this.name;
  }

  public findPasswordHash() {
    return this.name;
  }

  public findUserByEmailVerificationToken() {
    return this.name;
  }

  public findUserByResetPasswordToken() {
    return this.name;
  }

  public findUserByServiceId() {
    return this.name;
  }

  public addEmail() {
    return this.name;
  }

  public removeEmail() {
    return this.name;
  }

  public verifyEmail() {
    return this.name;
  }

  public setUsername() {
    return this.name;
  }

  public setPassword() {
    return this.name;
  }

  public setService() {
    return this.name;
  }

  public unsetService() {
    return this.name;
  }

  public createSession() {
    return this.name;
  }

  public updateSession() {
    return this.name;
  }

  public invalidateSession() {
    return this.name;
  }

  public invalidateAllSessions() {
    return this.name;
  }

  public removeAllResetPasswordTokens() {
    return this.name;
  }

  public findSessionByToken() {
    return this.name;
  }

  public findSessionById() {
    return this.name;
  }

  public addEmailVerificationToken() {
    return this.name;
  }

  public addResetPasswordToken() {
    return this.name;
  }

  public setUserDeactivated() {
    return this.name;
  }

  public findUserByLoginToken() {
    return this.name;
  }

  public addLoginToken() {
    return this.name;
  }

  public removeAllLoginTokens() {
    return this.name;
  }
}

const databaseManager = new DatabaseManager({
  userStorage: new Database('userStorage'),
  sessionStorage: new Database('sessionStorage'),
});

describe('DatabaseManager configuration', () => {
  it('should throw if no configuration object specified', () => {
    expect(() => (databaseManager as any).validateConfiguration(null as any)).toThrow();
  });

  it('should throw if no userStorage specified', () => {
    expect(() =>
      (databaseManager as any).validateConfiguration({ sessionStorage: true })
    ).toThrow();
  });

  it('should throw if no sessionStorage specified', () => {
    expect(() => (databaseManager as any).validateConfiguration({ userStorage: true })).toThrow();
  });

  it('should not throw if sessionStorage specified', () => {
    expect(() =>
      (databaseManager as any).validateConfiguration({
        userStorage: true,
        sessionStorage: true,
      })
    ).not.toThrow();
  });
});

describe('DatabaseManager', () => {
  it('createUser should be called on userStorage', () => {
    expect(databaseManager.createUser({ email: 'test', password: 'test' })).toBe('userStorage');
  });

  it('findUserById should be called on userStorage', () => {
    expect(databaseManager.findUserById('userId')).toBe('userStorage');
  });

  it('findUserByEmail should be called on userStorage', () => {
    expect(databaseManager.findUserByEmail('email')).toBe('userStorage');
  });

  it('findUserByUsername should be called on userStorage', () => {
    expect(databaseManager.findUserByUsername('username')).toBe('userStorage');
  });

  it('findPasswordHash should be called on userStorage', () => {
    expect(databaseManager.findPasswordHash('userId')).toBe('userStorage');
  });

  it('findUserByEmailVerificationToken should be called on userStorage', () => {
    expect(databaseManager.findUserByEmailVerificationToken('token')).toBe('userStorage');
  });

  it('findUserByResetPasswordToken should be called on userStorage', () => {
    expect(databaseManager.findUserByResetPasswordToken('token')).toBe('userStorage');
  });

  it('findUserByServiceId should be called on userStorage', () => {
    expect(databaseManager.findUserByServiceId('serviceName', 'serviceId')).toBe('userStorage');
  });

  it('addEmail should be called on userStorage', () => {
    expect(databaseManager.addEmail('userId', 'email', false)).toBe('userStorage');
  });

  it('removeEmail should be called on userStorage', () => {
    expect(databaseManager.removeEmail('userId', 'email')).toBe('userStorage');
  });

  it('verifyEmail should be called on userStorage', () => {
    expect(databaseManager.verifyEmail('userId', 'email')).toBe('userStorage');
  });

  it('setUsername should be called on userStorage', () => {
    expect(databaseManager.setUsername('userId', 'username')).toBe('userStorage');
  });

  it('setPassword should be called on userStorage', () => {
    expect(databaseManager.setPassword('userId', 'password')).toBe('userStorage');
  });

  it('setService should be called on userStorage', () => {
    expect(databaseManager.setService('userId', 'serviceName', {})).toBe('userStorage');
  });

  it('unsetService should be called on userStorage', () => {
    expect(databaseManager.unsetService('userId', 'serviceName')).toBe('userStorage');
  });

  it('createSession should be called on sessionStorage', () => {
    expect(databaseManager.createSession('userId', 'token', {})).toBe('sessionStorage');
  });

  it('updateSession should be called on sessionStorage', () => {
    expect(databaseManager.updateSession('sessionId', {})).toBe('sessionStorage');
  });

  it('invalidateSession should be called on sessionStorage', () => {
    expect(databaseManager.invalidateSession('sessionId')).toBe('sessionStorage');
  });

  it('invalidateAllSessions should be called on sessionStorage', () => {
    expect(databaseManager.invalidateAllSessions('userId')).toBe('sessionStorage');
  });

  it('findSessionByToken should be called on sessionStorage', () => {
    expect(databaseManager.findSessionByToken('token')).toBe('sessionStorage');
  });

  it('findSessionById should be called on sessionStorage', () => {
    expect(databaseManager.findSessionById('sessionId')).toBe('sessionStorage');
  });

  it('addEmailVerificationToken should be called on sessionStorage', () => {
    expect(databaseManager.addEmailVerificationToken('userId', 'email', 'token')).toBe(
      'userStorage'
    );
  });

  it('addResetPasswordToken should be called on sessionStorage', () => {
    expect(databaseManager.addResetPasswordToken('userId', 'email', 'token', 'reason')).toBe(
      'userStorage'
    );
  });

  it('setUserDeactivated should be called on sessionStorage', () => {
    expect(databaseManager.setUserDeactivated('userId', true)).toBe('userStorage');
  });

  it('removeAllResetPasswordTokens should be called on userStorage', () => {
    expect(databaseManager.removeAllResetPasswordTokens('userId')).toBe('userStorage');
  });

  it('addLoginToken should be called on userStorage', () => {
    expect(databaseManager.addLoginToken('userId', 'email', 'token')).toBe('userStorage');
  });

  it('removeAllLoginTokens should be called on userStorage', () => {
    expect(databaseManager.removeAllLoginTokens('userId')).toBe('userStorage');
  });

  it('findUserByLoginToken should be called on userStorage', () => {
    expect(databaseManager.findUserByLoginToken('token')).toBe('userStorage');
  });
});
