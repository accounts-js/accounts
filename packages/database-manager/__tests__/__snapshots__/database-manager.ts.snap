// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatabaseManager configuration without MFA createMfaLoginAttempt should throw error 1`] = `"No mfaLoginAttemptsStorage defined for manager"`;

exports[`DatabaseManager configuration without MFA getMfaLoginAttempt should throw error 1`] = `"No mfaLoginAttemptsStorage defined for manager"`;

exports[`DatabaseManager configuration without MFA removeMfaLoginAttempt should throw error 1`] = `"No mfaLoginAttemptsStorage defined for manager"`;
